db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/CastNode.java

```diff
@@ -320,6 +324,20 @@ else if (this.targetCharType == Types.VARCHAR)
 
         bindCastNodeOnly();
 
+        if (getTypeId().isCharOrVarChar()) {
+            if (requestedStringLength != -1 && !sourceCTI.isCharOrVarChar()) {
+                throw StandardException.newException(
+                        SQLState.LANG_INVALID_CAST_TO_CHAR_WITH_LENGTH_NOT_FROM_CHAR,
+                        sourceCTI.getSQLTypeName(),
+                        getTypeId().getSQLTypeName());
+            }
+            if (dateToStringFormat != -1 && !sourceCTI.isDateTimeTimeStampTypeID()) {
+                throw StandardException.newException(
+                        SQLState.LANG_INVALID_CAST_TO_CHAR_WITH_FORMAT_NOT_FROM_DATE
+                );
+            }
```

Comment:
I'm not sure if this is always fine. I wanted to do check here, but then there are cases where `sourceCTI` is null.

Reply:
Do you know when that can be null?
In the meantime, I added a nullcheck

Reply:
I moved line 367 and 368 out, then added a check like `if (sourceJDBCTypeId == Types.TIMESTAMP && stringFormat >= 0)` then throw (because DB2 doesn’t accept datetime format for timestamp), and run tests, then there will be NPE.